wl/server/HandleCommand.java

```diff
@@ -260,159 +301,166 @@ public void handleCmdSubmit() throws Exception {
-			for (int i = 0; i < nrDirs; ++i) {
-				String n = ServerUtils.readLine(in);
-				if (n.startsWith("/")) n = n.substring(1);
-				ServerUtils.checkNameValid(n, true);
-				dirnames[i] = new File(tempDir, n);
-				dirnames[i].mkdirs();
+		ServerUtils.semaphoreRW(cmd[1], () -> {
+			if (!admin) {
+				File f = new File("metadata", cmd[1] + ".maintain");
+				if (f.exists()) {
+					String originalUploader =
+					    Utils.readProfile(f, cmd[1]).get("uploader").value(locale);
+					if (!username.equals(originalUploader))
+						throw new ServerUtils.WLProtocolException(
+						    "You can not overwrite another person's (" + originalUploader +
+						    ") existing add-on");
+				}
 			}
+			File tempDir = Utils.createTempDir();
 
-			long totalSize = 0;
-			for (int i = 0; i < nrDirs; ++i) {
-				final int nrFiles = Integer.valueOf(ServerUtils.readLine(in));
-				if (nrFiles < 0 || nrFiles > 1000)
+			try {
+				final int nrDirs = Integer.valueOf(ServerUtils.readLine(in));
+				if (nrDirs < 0 || nrDirs > 1000)
 					throw new ServerUtils.WLProtocolException(
-					    "File count limit of 1000 exceeded. "
+					    "Directory count limit of 1000 exceeded. "
 					    + "If you really want to submit such a large add-on, "
 					    + "please contact the Widelands Development Team.");
-				for (int j = 0; j < nrFiles; ++j) {
-					final String filename = ServerUtils.readLine(in);
-					ServerUtils.checkNameValid(filename, false);
-					final String checksum = ServerUtils.readLine(in);
-					final long size = Long.valueOf(ServerUtils.readLine(in));
-					totalSize += size;
-					if (totalSize < 0 || totalSize > 200 * 1000 * 1000)
+				File[] dirnames = new File[nrDirs];
+				for (int i = 0; i < nrDirs; ++i) {
+					String n = ServerUtils.readLine(in);
+					if (n.startsWith("/")) n = n.substring(1);
+					ServerUtils.checkNameValid(n, true);
+					dirnames[i] = new File(tempDir, n);
+					dirnames[i].mkdirs();
+				}
+
+				long totalSize = 0;
+				for (int i = 0; i < nrDirs; ++i) {
+					final int nrFiles = Integer.valueOf(ServerUtils.readLine(in));
+					if (nrFiles < 0 || nrFiles > 1000)
 						throw new ServerUtils.WLProtocolException(
-						    "Filesize limit of 200 MB exceeded. "
+						    "File count limit of 1000 exceeded. "
 						    + "If you really want to submit such a large add-on, "
 						    + "please contact the Widelands Development Team.");
-					File file = new File(dirnames[i], filename);
-					PrintStream stream = new PrintStream(file);
-					for (long l = 0; l < size; ++l) {
-						int b = in.read();
-						if (b < 0)
+					for (int j = 0; j < nrFiles; ++j) {
+						final String filename = ServerUtils.readLine(in);
+						ServerUtils.checkNameValid(filename, false);
+						final String checksum = ServerUtils.readLine(in);
+						final long size = Long.valueOf(ServerUtils.readLine(in));
+						totalSize += size;
+						if (totalSize < 0 || totalSize > 200 * 1000 * 1000)
+							throw new ServerUtils.WLProtocolException(
```

Comment:
Do you clean up the directory, if an addon is rejected due to its size or other errors?

Reply:
Yes, this is encapsulated in a huge `try` block that stretches from line 319 to 460. The `catch` block cleans up in line 461 if anything at all goes wrong.